File: frontend/src/types/workflow.ts
/** Runtime-facing types mirrored from /contracts and guarded by contract tests. */

export const stageNames = [
  'data_fetch',
  'data_interpret',
  'chart_generate',
  'chapter_write',
  'report_fusion',
] as const

export const stageStatuses = [
  'pending',
  'running',
  'waiting_review',
  'approved',
  'rejected',
  'completed',
  'failed',
  'cancelled',
] as const

export const reviewActions = [
  'approve',
  'accept_recommendation',
  'accept_with_risks',
  'customize',
  'revise',
  'regenerate',
  'cancel',
] as const

export type StageName = (typeof stageNames)[number]
export type StageStatus = (typeof stageStatuses)[number]
export type ReviewAction = (typeof reviewActions)[number]

export type RiskSeverity = 'info' | 'warning' | 'high' | 'critical'
export type RiskDisposition = 'advisory' | 'acknowledgement_required' | 'hard_block'
export type DecisionStatus =
  | 'not_required'
  | 'awaiting_user'
  | 'accepted_recommendation'
  | 'accepted_with_risks'
  | 'customized'
  | 'cancelled'

export interface RiskNotice {
  risk_code: string
  stage: string
  severity: RiskSeverity
  disposition: RiskDisposition
  title: string
  detail: string
  affected_ids: string[]
  recommendation: string
  consequence: string
  can_override: boolean
}

export interface ChartCandidateResult {
  candidate_id: string
  title: string
  chart_type: string
  status: string
  recommended_chapter_id: string | null
  alternative_chapter_ids: string[]
  priority: number
  evidence_ids: string[]
  risk_notices: RiskNotice[]
  conflict_group_id: string | null
  chart_id: string | null
  suppression_reason: string | null
}

export interface ConflictGroup {
  conflict_group_id: string
  candidate_ids: string[]
  recommended_candidate_id: string
  reason: string
  risk_if_keep_all: string
}

export interface DecisionPackage {
  decision_id: string
  run_id: string
  stage: string
  revision: number
  all_candidates: ChartCandidateResult[]
  recommended_selection: string[]
  conflict_groups: ConflictGroup[]
  risk_notices: RiskNotice[]
  blocking_risk_codes: string[]
  acknowledgement_required_codes: string[]
  decision_status: DecisionStatus
}

export interface UserDecision {
  decision_id: string
  run_id: string
  owner_id: string
  stage: string
  action: ReviewAction
  selected_chart_ids: string[]
  excluded_chart_ids: string[]
  placement_overrides: Record<string, string>
  accepted_risk_codes: string[]
  release_mode: 'formal' | 'draft_with_warnings'
  comment: string | null
  expected_revision: number
  risk_snapshot_sha256: string
}

export interface ArtifactRef {
  artifact_id: string
  kind: string
  uri: string
  checksum: string | null
  revision: number
}

export interface StageResult {
  stage: StageName
  status: StageStatus
  revision: number
  data: Record<string, unknown>
  artifacts: ArtifactRef[]
  evidence_sources: string[]
  error: string | null
}

export const p0ChartTypes = ['line', 'bar', 'pie', 'radar', 'industry_chain'] as const
export const p1ChartTypes = [
  'combo',
  'area',
  'scatter',
  'bubble',
  'heatmap',
  'boxplot',
  'treemap',
] as const
export const chartTypes = [...p0ChartTypes, ...p1ChartTypes] as const
export type P0ChartType = (typeof p0ChartTypes)[number]
export type P1ChartType = (typeof p1ChartTypes)[number]
export type ChartType = (typeof chartTypes)[number]
export type ChartVariant =
  | 'line'
  | 'vertical'
  | 'horizontal'
  | 'grouped'
  | 'stacked'
  | 'pie'
  | 'radar'
  | 'graph'
  | 'combo'
  | 'area'
  | 'scatter'
  | 'bubble'
  | 'heatmap'
  | 'boxplot'
  | 'treemap'

export interface ChartReference {
  chart_id: string
  title: string
  chart_type: ChartType
  status: 'planned' | 'ready'
  evidence_ids: string[]
  artifact_id: string | null
  candidate_status?: string | null
}

export interface ChartSpec {
  chart_id: string
  title: string
  chart_type: ChartType
  variant: ChartVariant
  option: Record<string, unknown>
  evidence_ids: string[]
  data_fingerprint: string
  dedupe_key: string
}

export interface ChartGenerationResult {
  charts: ChartReference[]
  chart_specs: ChartSpec[]
  suppressed_candidates: Array<{
    title: string
    reason_code: string
    reason: string
    evidence_ids: string[]
  }>
  quality: {
    passed: boolean
    ready_count: number
    suppressed_count: number
    issues: string[]
  }
}

export const reportFormats = ['markdown', 'html', 'pdf'] as const
export type ReportFormat = (typeof reportFormats)[number]

export interface ReportArtifactManifestEntry {
  artifact_id: string
  kind: 'report_markdown' | 'report_html' | 'report_pdf' | 'artifact_manifest'
  uri: string
  sha256: string
  size_bytes: number
}

export interface ReportFusionResult {
  report_id: string
  title: string
  industry_topic: string
  research_as_of: string
  generated_at: string
  tone: 'professional' | 'plain_language'
  formats: ReportFormat[]
  source_revisions: Array<{
    stage: 'data_interpret' | 'chart_generate' | 'chapter_write'
    revision: number
  }>
  included_chart_ids: string[]
  artifacts: ReportArtifactManifestEntry[]
  quality: {
    passed: boolean
    chapter_count: number
    section_count: number
    included_chart_count: number
    evidence_coverage: number
    issues: string[]
  }
  release_mode: 'formal' | 'draft_with_warnings'
  formal_eligible: boolean
  draft_eligible: boolean
  acknowledged_risks: string[]
  unresolved_risks: string[]
}

export interface WorkflowState {
  project_id: string
  run_id: string
  current_stage: StageName
  status: StageStatus
  revision: number
  stage_results: Partial<Record<StageName, StageResult>>
  created_at: string
  updated_at: string
}

export interface RunCreateRequest {
  project_id: string
  run_id?: string | null
  input_data: Record<string, unknown>
  review_stages: StageName[]
}

export interface ReviewRequest {
  run_id: string
  stage: StageName
  action: ReviewAction
  expected_revision: number
  comment: string | null
  edited_data: Record<string, unknown> | null
  accepted_risk_codes?: string[]
  release_mode?: 'formal' | 'draft_with_warnings'
  selected_chart_ids?: string[]
  placement_overrides?: Record<string, string>
}

File: backend/app/schemas/workflow.py
from __future__ import annotations
from datetime import datetime
from typing import Any, Literal
from pydantic import BaseModel, Field

StageName = Literal[
    "data_fetch",
    "data_interpret",
    "chart_generate",
    "chapter_write",
    "report_fusion",
]

STAGE_ORDER: list[StageName] = [
    "data_fetch",
    "data_interpret",
    "chart_generate",
    "chapter_write",
    "report_fusion",
]

StageStatus = Literal[
    "pending",
    "running",
    "waiting_review",
    "approved",
    "rejected",
    "completed",
    "failed",
    "cancelled",
]

ReviewAction = Literal[
    "approve",
    "accept_recommendation",
    "accept_with_risks",
    "customize",
    "revise",
    "regenerate",
    "cancel",
    "direct_edit",
]

AnalysisDepth = Literal["overview", "standard", "deep"]
RiskPreference = Literal["conservative", "balanced", "aggressive"]
ReleaseMode = Literal["formal", "draft_with_warnings"]


class ArtifactRef(BaseModel):
    artifact_id: str
    kind: str
    uri: str
    checksum: str | None = None
    revision: int = 1


class StageResult(BaseModel):
    stage: StageName
    status: StageStatus = "pending"
    revision: int = 1
    data: dict[str, Any] = Field(default_factory=dict)
    artifacts: list[ArtifactRef] = Field(default_factory=list)
    evidence_sources: list[str] = Field(default_factory=list)
    error: str | None = None


class AgentTraceEvent(BaseModel):
    id: str
    timestamp: str = Field(default_factory=lambda: datetime.now().isoformat())
    stage: StageName
    stage_label: str
    event_type: str = "info"  # agent_start | tool_call | llm_thought | artifact_created | stage_completed | error | info
    message: str
    tool: str | None = None
    details: dict[str, Any] = Field(default_factory=dict)


class WorkflowState(BaseModel):
    project_id: str
    run_id: str
    current_stage: StageName = "data_fetch"
    status: StageStatus = "pending"
    revision: int = 1
    stage_results: dict[StageName, StageResult] = Field(default_factory=dict)
    created_at: str = Field(default_factory=lambda: datetime.now().isoformat())
    updated_at: str = Field(default_factory=lambda: datetime.now().isoformat())


class RunSummary(BaseModel):
    run_id: str
    project_id: str
    title: str
    current_stage: StageName
    status: StageStatus
    revision: int
    created_at: str
    updated_at: str
    artifact_count: int = 0
    report_available: bool = False


class RunListResponse(BaseModel):
    total: int
    offset: int
    limit: int
    items: list[RunSummary]


class RevisionSummary(BaseModel):
    revision: int
    status: StageStatus
    current_stage: StageName
    updated_at: str


class RevisionListResponse(BaseModel):
    run_id: str
    current_revision: int
    revisions: list[RevisionSummary]


class DataFetchOptions(BaseModel):
    keywords: list[str] | None = None
    industry_scope: list[str] | None = None
    time_range: list[str] | None = None
    data_sources: list[str] | None = None
    metrics: list[str] | None = None


class ResearchBrief(BaseModel):
    geography: str | None = None
    time_range: str | None = None
    included_topics: list[str] | None = None
    excluded_topics: list[str] | None = None
    focus_companies: list[str] | None = None
    report_depth: Literal["brief", "standard", "deep"] | None = None


class ChartGenerationOptions(BaseModel):
    chart_mode: str | None = None
    chart_type: str | None = None
    requested_chart_count: int | None = None
    requested_chart_types: list[str] | None = None
    user_priority: bool | None = None
    allow_multiple_charts_per_dataset: bool | None = None
    bar_variant: Literal["vertical", "horizontal", "grouped", "stacked"] | None = None
    metric_ids: list[str] | None = None
    title: str | None = None
    color_theme: str | None = None
    emphasis: str | None = None


class ResearchInput(BaseModel):
    industry_topic: str
    market_scope: list[str] = Field(default_factory=lambda: ["中国 A 股", "港股", "美股", "中国 B 股"])
    security_types: list[str] = Field(default_factory=lambda: ["股票", "债券", "基金", "期货", "指数"])
    reporting_currency: str = "CNY"
    research_as_of: str = Field(default_factory=lambda: datetime.now().strftime("%Y-%m-%d"))
    focus_questions: list[str] = Field(default_factory=list)
    data_fetch_options: DataFetchOptions | None = None
    analysis_depth: AnalysisDepth = "standard"
    risk_preference: RiskPreference = "balanced"
    research_brief: ResearchBrief | None = None
    chart_generate_options: ChartGenerationOptions | None = None


class RunCreateRequest(BaseModel):
    project_id: str
    input_data: ResearchInput
    review_stages: list[StageName] = Field(
        default_factory=lambda: [
            "data_fetch",
            "data_interpret",
            "chart_generate",
            "chapter_write",
            "report_fusion",
        ]
    )


class ReviewRequest(BaseModel):
    run_id: str
    stage: StageName
    action: ReviewAction
    expected_revision: int
    comment: str | None = None
    edited_data: dict[str, Any] | None = None
    accepted_risk_codes: list[str] | None = None
    release_mode: ReleaseMode | None = None
    selected_chart_ids: list[str] | None = None
    decision_id: str | None = None
    risk_snapshot_sha256: str | None = None
